ERCS/erc-7786.md

```diff
@@ -0,0 +1,222 @@
+---
+eip: 7786
+title: Cross-Chain Messaging Gateway
+description: An interface for contracts to send and receive cross-chain messages.
+author: Francisco Giordano (@frangio), Hadrien Croubois (@Amxx), Ernesto Garcia (@ernestognw), CJ Cobb (@cjcobb23)
+discussions-to: https://ethereum-magicians.org/t/erc-7786-cross-chain-messaging-gateway/21374
+status: Draft
+type: Standards Track
+category: ERC
+created: 2024-10-14
+---
+
+## Abstract
+
+This proposal describes an interface, and the corresponding workflow, for smart contracts to send arbitrary data through cross-chain messaging systems. The end goal of this proposal is to have all such messaging protocols accessible via this interface (natively or using "adapters") to improve their composability and interoperability. That would allow a new class of cross-chain native smart contracts to emerge while reducing vendor lock-in. This proposal is modular by design, allowing users to leverage bridge-specific features through attributes while providing simple "universal" access to the simple feature of "just getting a simple message through".
+
+## Motivation
+
+Cross-chain messaging systems (or bridges) allow communication between smart contracts deployed on different blockchains. There is a large diversity of such systems with multiple degrees of decentralization, with various components, that implement different interfaces, and provide different guarantees to the users.
+
+Because almost every protocol implementing a different workflow, using a specific interface, portability between bridges is basically impossible. This also forbid the development of generic contracts that rely on cross chain communication.
+
+The objective of the ERC is to provide a standard interface, and a corresponding workflow, for performing cross-chain communication between contracts. Existing cross-chain communication protocols, that do not nativelly implement this interface, should be able to adopt it using adapter gateway contracts.
+
```

Comment:
There have been quite a few standards of these nature proposed, and none of them have seemingly achieved any traction.

Here is an overview of the state of these standards: https://github.com/0xFableOrg/xchain/blob/master/README.md#messaging-formats--interfaces

On the ERC front only, we have:
- [ERC-5164: Cross-Chain Execution](https://eips.ethereum.org/EIPS/eip-5164)
- [ERC-6170: Cross-Chain Messaging Interface](https://eips.ethereum.org/EIPS/eip-6170)
 - [ERC-7533: Public Cross Port](https://eips.ethereum.org/EIPS/eip-7533)
 
Do we need another one? What makes this one different / necessary? And most importantly, though I'm not sure that's in scope for the EIP motivation section: what will make it so that this one gets actually used?

Reply:
Thanks for the feedback! We should definitely address these points in the ERC. My views are:

ERC-7533 is very low level and seems to mandate a lot of details about the protocol. The goal of this new ERC is to be a viable interface for any underlying cross-chain messaging protocol.

ERC-5164's main drawback is lack of support for non-EVM chains, since it relies on the EVM's chain id. ERC-6170 technically supports any chain id format but doesn't specify which one, so it doesn't achieve portability. This new ERC mandates the use of CAIP-2 and CAIP-10 identifiers.

ERC-5164 also lacks the ability to expose other features/parameters of the underlying protocol such as setting a minimum gas limit and attaching tokens. ERC-6170 has a separate "bytes data" argument that could technically be used for these things but does not specify a format or way to query whether they're supported, so it would struggle with portability as is. This new ERC has an extensible mechanism for attributes that can expose more advanced bridge features, with a viable path to standardizing the different kinds of attributes.

> what will make it so that this one gets actually used?

The theory is that if the interface is simple and unopinionated yet expressive enough, it should be a no-brainer to adopt for the purpose of increasing interoperability and composability. We're working with Axelar and actively reaching out to developers of other cross-chain messaging protocols to get a critical mass.

Reply:
@norswap I'd be interested to hear your thoughts on the spec!

Reply:
In addition to what @frangio said, something else I would like to add is that neither ERC 5164 nor ERC 6170 standardize the interface that apps use to verify messages with the gateway/interop protocol. Which is a major component of any cross chain app, and without standardizing that layer, it is impossible to build a generic cross chain app, since the app is still tightly coupled with the verification of messages.

Reply:
@frangio Thanks for the answer, that all makes sense.

@cjcobb23 I'm not sure I fully understand — EIP6170 has a `receiveMessage`. The difference seems to be that theirs live in their inbox (gateway), which then will call the app (as presumably the calldata is a message to the app). If I understand correctly, that's the same thing as the "active" mode in this EIP (with the difference is that the gateway function is unspecified and calls the `receiveMessage` on the destination contract instead of another function of the app's choosing).

While the passive mode is interesting, I don't think it really enables more use cases? The passive mode allows you to abstract the gateway (but they still need to be verified). This can equivalently be done in EIP6170 by having their inbox `receiveMessage` call a function in the app that checks the `msg.sender` (who then similarly checks against a list of known inboxes).

Reply:
yes I was talking about passive mode. But even in active mode, does EIP 6170 standardize the interface between the gateway and the dApp the message is sent to? What interface must the dApp implement to receive cross chain messages? Is the `message_` field just calldata, and is the original sender and source chain propagated to the dApp and if so how? I think these things are all important to spell out.